docs/learn/output_parsers.md

```diff
@@ -107,16 +160,270 @@ else:
+        """)
+    def recommend_book(genre: str):
+        ...
+    
+    result = recommend_book("science fiction")
+    print(f"Title: {result['title']}")
+    print(f"Author: {result['author']}")
+    ```
+
+2. **Handle Parsing Errors**: Always implement error handling in your parsers. LLMs may occasionally produce outputs that don't conform to the expected structure.
+
+    ```python
+    import xml.etree.ElementTree as ET
+    from mirascope.core import anthropic, prompt_template
+    
+    def parse_book_recommendation(response: anthropic.AnthropicCallResponse) -> dict:
+        try:
+            root = ET.fromstring(response.content)
+            required_elements = ["title", "author", "year", "summary"]
+            if not all(root.find(elem) is not None for elem in required_elements):
+                raise ValueError("Missing required elements in the XML")
+            return {
+                "title": root.find("title").text,
+                "author": root.find("author").text,
+                "year": int(root.find("year").text),
+                "summary": root.find("summary").text
+            }
+        except ET.ParseError:
+            print("Failed to parse XML from the response")
+            return {}
+        except ValueError as e:
+            print(f"Invalid data structure: {e}")
+            return {}
+    
+    @anthropic.call("claude-3-5-sonnet-20240620", output_parser=parse_book_recommendation)
+    @prompt_template("Recommend a {genre} book with title, author, year, and summary in XML format.")
+    def recommend_book(genre: str):
+        ...
+    
+    result = recommend_book("mystery")
+    if result:
+        print(f"Title: {result.get('title', 'N/A')}")
+        print(f"Author: {result.get('author', 'N/A')}")
+    else:
+        print("Failed to get a valid recommendation")
+    ```
+
+3. **Gradual Refinement**: Start with simple parsers and gradually increase complexity as you refine your prompts and understand the model's output patterns.
+
+    ```python
+    import xml.etree.ElementTree as ET
+    from mirascope.core import anthropic, prompt_template
+    
+    # Simple parser
+    def parse_title_author(response: anthropic.AnthropicCallResponse) -> dict:
+        root = ET.fromstring(response.content)
+        return {
+            "title": root.find("title").text,
+            "author": root.find("author").text
+        }
+    
+    # More complex parser
+    def parse_book_details(response: anthropic.AnthropicCallResponse) -> dict:
+        root = ET.fromstring(response.content)
+        return {
+            "title": root.find("title").text,
+            "author": root.find("author").text,
+            "year": int(root.find("year").text),
+            "summary": root.find("summary").text,
+            "genres": [genre.text for genre in root.findall("genre")]
+        }
+    
+    @anthropic.call("claude-3-5-sonnet-20240620", output_parser=parse_title_author)
```

Comment:
should we colocate this code with the corresponding parser rather than colocating the parsers and methods separately?